lib/async_hooks.js

```diff
@@ -223,80 +223,67 @@ const storageHook = createHook({
 class AsyncLocalStorage {
   constructor() {
     this.kResourceStore = Symbol('kResourceStore');
-    this.enabled = false;
   }
 
   disable() {
-    if (this.enabled) {
-      this.enabled = false;
-      // If this.enabled, the instance must be in storageList
-      storageList.splice(storageList.indexOf(this), 1);
-      if (storageList.length === 0) {
-        storageHook.disable();
-      }
+    const resource = executionAsyncResource();
+    resource[this.kResourceStore] = undefined;
+
+    // If this.enabled, the instance must be in storageList
+    storageList.splice(storageList.indexOf(this), 1);
+    if (storageList.length === 0) {
+      storageHook.disable();
     }
   }
 
   // Propagate the context from a parent resource to a child one
   _propagate(resource, triggerResource) {
     const store = triggerResource[this.kResourceStore];
-    if (this.enabled) {
-      resource[this.kResourceStore] = store;
-    }
+    resource[this.kResourceStore] = store;
   }
 
   _enter(store) {
-    if (!this.enabled) {
-      this.enabled = true;
+    if (storageList.indexOf(this) === -1) {
       storageList.push(this);
       storageHook.enable();
     }
     const resource = executionAsyncResource();
     resource[this.kResourceStore] = store;
   }
 
-  _exit() {
-    const resource = executionAsyncResource();
-    if (resource) {
-      resource[this.kResourceStore] = undefined;
-    }
-  }
-
   runSyncAndReturn(store, callback, ...args) {
-    this._enter(store);
-    try {
+    const resource = new AsyncResource('AsyncLocalStorage');
+    return resource.runInAsyncScope(() => {
+      this._enter(store);
       return callback(...args);
-    } finally {
-      this._exit();
-    }
+    });
   }
 
   exitSyncAndReturn(callback, ...args) {
-    this.enabled = false;
-    try {
+    const resource = new AsyncResource('AsyncLocalStorage');
+    return resource.runInAsyncScope(() => {
```

Comment:
What is the performance impact of this change?

Reply:
I also share this concern. It might be a good idea to run `async-resource-vs-destroy.js` benchmark before and after these changes.

Reply:
```
./node benchmark/compare.js --new ./node --old ./node-master --runs 10 --filter async-resource async_hooks | Rscript benchmark/compare.R
[00:10:20|% 100| 1/1 files | 20/20 runs | 6/6 configs]: Done
                                                                                                                                       confidence improvement accuracy (*)   (**)  (***)
 async_hooks/async-resource-vs-destroy.js n=1000000 asyncMethod='async' type='async-local-storage' benchmarker='test-double-http'                    -1.34 %       ±1.84% ±2.54% ±3.50%
 async_hooks/async-resource-vs-destroy.js n=1000000 asyncMethod='async' type='async-resource' benchmarker='test-double-http'                          0.63 %       ±1.10% ±1.51% ±2.06%
 async_hooks/async-resource-vs-destroy.js n=1000000 asyncMethod='async' type='destroy' benchmarker='test-double-http'                                -0.09 %       ±1.70% ±2.34% ±3.19%
 async_hooks/async-resource-vs-destroy.js n=1000000 asyncMethod='callbacks' type='async-local-storage' benchmarker='test-double-http'                -1.02 %       ±1.59% ±2.19% ±3.00%
 async_hooks/async-resource-vs-destroy.js n=1000000 asyncMethod='callbacks' type='async-resource' benchmarker='test-double-http'                     -0.11 %       ±1.19% ±1.63% ±2.22%
 async_hooks/async-resource-vs-destroy.js n=1000000 asyncMethod='callbacks' type='destroy' benchmarker='test-double-http'                            -0.76 %       ±1.61% ±2.24% ±3.12%
```

Reply:
Basically, the difference is a rounding error.

Reply:
Marking as resolved. Feel free to revert if you disagree.